File: v103/src/get_rating.py
import json
import os
import dicttoxml
import xmltodict
import requests
import logging
from ast import literal_eval
from datetime import date
LOGGER = logging.getLogger()
LOGGER.setLevel(logging.INFO)

INTERNAL_ERROR_MESSAGE = "Internal Error."

def handler(event, context):
    LOGGER.info("Event: %s", json.dumps(event))
    
    start = """<?xml version="1.0" encoding="utf-8"?><soap12:Envelope \
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" \
        xmlns:xsd="http://www.w3.org/2001/XMLSchema" \
            xmlns:soap12="http://www.w3.org/2003/05/soap-envelope"><soap12:Body> \
    <GetRating xmlns="http://tempuri.org/">\
      <RatingParam><RatingInput>"""
    rating_input = """"""
    commodity_input = """"""
    mid = """</RatingInput><CommodityInput><CommodityInput>"""
    end = """</CommodityInput></CommodityInput></RatingParam></GetRating></soap12:Body></soap12:Envelope>"""
    payload = start+rating_input+mid+commodity_input+end
    LOGGER.info("Payload xml data is : %s", json.dumps(payload))

    try:
        url = "https://wttest.omnilogistics.com/WTKServices/getrating.asmx"
    except Exception as url_error:
        LOGGER.exception("Environment variable URL not set.")
        raise EnvironmentVariableError(json.dumps({"httpStatus": 501, "message": INTERNAL_ERROR_MESSAGE})) from url_error
    try:
        req = requests.post(url, headers = {'Content-Type': 'text/xml; charset=utf-8'},data = payload)
        response = req.text
        LOGGER.info("Response is : %s", json.dumps(response))
    except Exception as wt_error:
        LOGGER.exception("GetRatingApiError: %s", json.dumps(wt_error))
        raise GetRatingApiError(json.dumps({"httpStatus": 400, "message": "WorldTrack Get Rating Api Error"})) from wt_error

def modify_object_keys(array):
    new_array = []
    for obj in array:
        new_obj = {}
        for key in obj:
            new_key = key.replace(" ","")
            new_obj[new_key] = obj[key]
        new_array.append(new_obj)
    return new_array

def validate_input(event):
    if not "enhancedAuthContext" in event or "customerId" not in event["enhancedAuthContext"]:
        raise InputError(json.dumps({"httpStatus": 400, "message": "CustomerId not found."}))
    client_data = ['Service Level','Ready Date']
    if not "body" in event or not "oShipData" in event["body"] or not set(client_data).issubset(event["body"]["oShipData"]):
        raise InputError(json.dumps({"httpStatus": 400, "message": "One/All of: Service Level, Ready Date parameters are missing in the request body oShipData."}))
    return event["enhancedAuthContext"]["customerId"]

class InputError(Exception):
    pass
class HandlerError(Exception):
    pass
class GetShipmentLineListError(Exception):
    pass
class GetReferenceListError(Exception):
    pass
class GetAccessorialListError(Exception):
    pass
class WtBolApiError(Exception):
    pass
class DataTransformError(Exception):
    pass
class EnvironmentVariableError(Exception):
    pass
class AirtrakShipmentApiError(Exception):
    pass
class GetServiceLevelError(Exception):
    pass
class ReadyDateTimeError(Exception):
    pass

File: v103/src/get_rating.js
const AWS = require("aws-sdk");
const Joi = require("joi");
const axios = require("axios");
const { convert } = require("xmlbuilder2");

const replaceNull = Joi.allow(null).empty("").default(null);
const CommodityInputValidation = {
  CommodityClass: Joi.number().precision(2).concat(replaceNull),
  CommodityPieces: Joi.number().integer().concat(replaceNull),
  CommodityPieceType: Joi.string().concat(replaceNull),
  CommodityWeightPerPiece: Joi.number().integer().concat(replaceNull),
  CommodityWeight: Joi.number().integer().concat(replaceNull),
  CommodityLength: Joi.number().integer().concat(replaceNull),
  CommodityWidth: Joi.number().integer().concat(replaceNull),
  CommodityHeight: Joi.number().integer().concat(replaceNull),
  CommodityHazmat: Joi.string().concat(replaceNull),
};
const eventValidation = Joi.object().keys({
  RatingInput: Joi.object()
    .keys({
      RequestID: Joi.number().integer(),
      OriginCountry: Joi.string(),
      OriginCity: Joi.string(),
      OriginState: Joi.string(),
      OriginZip: Joi.number().integer().required(),
      DestinationCountry: Joi.string(),
      DestinationCity: Joi.string(),
      DestinationState: Joi.string(),
      DestinationZip: Joi.number().integer().required(),
      ShipmentTerms: Joi.string(),
      PickupDate: Joi.date().iso().greater("now").required(),
      PickupTime: Joi.date().iso().greater("now").required(),
      PickupLocationCloseTime: Joi.date().iso().greater("now").required(),
    })
    .required(),
  CommodityInput: Joi.object()
    .keys({
      CommodityInput: Joi.alternatives(
        Joi.object().keys(CommodityInputValidation),
        Joi.array().items(CommodityInputValidation)
      ).required(),
    })
    .required(),
});

module.exports.handler = async (event, context, callback) => {
  const LiabilityType = "LL",
    WebTrakUserID = "biztest";

  const { error, value } = eventValidation.validate(event.body);
  if (error) {
    return errorMsg(400, "Please provide all required fields.", error.details);
  }
  const apiKey = event.headers["x-api-key"];
  const eventBody = value;
  eventBody.RatingInput.LiabilityType = LiabilityType;
  eventBody.RatingInput.WebTrakUserID = WebTrakUserID;
  eventBody.RatingInput.PickupDate =
    event.body.RatingInput.PickupDate.toString();
  eventBody.RatingInput.PickupTime =
    event.body.RatingInput.PickupTime.toString();
  eventBody.RatingInput.PickupLocationCloseTime =
    event.body.RatingInput.PickupLocationCloseTime.toString();
  try {
    const customerData = await getCustomerId(apiKey);
    if (customerData == null) {
      return errorMsg(400, "Api key validation error");
    } else if (
      !customerData.hasOwnProperty("WebTrackId") ||
      customerData.WebTrackId == null
    ) {
      return errorMsg(400, "No valid WebTrackId");
    }
    eventBody.RatingInput.WebTrackId = customerData.WebTrackId;

    const postData = makeJsonToXml(eventBody);
    console.log(postData);
    // return {};
    const dataResponse = await getRating(postData);
    const dataObj = makeXmlToJson(dataResponse);
    return dataObj;
  } catch (error) {
    return errorMsg(400, "Something went wrong.", error);
  }
};

function errorMsg(code, message, error = null) {
  return {
    httpStatus: code,
    code,
    message,
    error,
  };
}

function makeJsonToXml(data) {
  return convert({
    "soap12:Envelope": {
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xmlns:xsd": "http://www.w3.org/2001/XMLSchema",
      "@xmlns:soap12": "http://www.w3.org/2003/05/soap-envelope",
      "soap12:Body": {
        GetRating: {
          "@xmlns": "http://tempuri.org/",
          RatingParam: data,
        },
      },
    },
  });
}

function makeXmlToJson(data) {
  let obj = convert(data, { format: "object" });
  return obj["soap:Envelope"]["soap:Body"].GetRatingResponse.GetRatingResult
    .RatingOutput;
}

function getCustomerId(ApiKey) {
  return new Promise(async (resolve, reject) => {
    const documentClient = new AWS.DynamoDB.DocumentClient({
      region: process.env.REGION,
    });
    const params = {
      TableName: process.env.TOKEN_VALIDATOR,
      FilterExpression: "#ApiKey = :ApiKey",
      ExpressionAttributeNames: { "#ApiKey": "ApiKey" },
      ExpressionAttributeValues: { ":ApiKey": ApiKey },
    };
    try {
      const response = await documentClient.scan(params).promise();
      if (response.Items && response.Items.length > 0) {
        resolve(response.Items[0]);
      } else {
        resolve(null);
      }
    } catch (e) {
      reject("getCustomerId Error: Request failed with status code 500");
    }
  });
}

function getRating(postData) {
  return new Promise(async (resolve, reject) => {
    try {
      const res = await axios.post(
        "https://worldtrak.omnilogistics.com/WTKServices/GetRating.asmx",
        postData,
        {
          headers: {
            Accept: "text/xml",
            "Content-Type": "text/xml; charset=utf-8",
          },
        }
      );
      if (res.status == 200) {
        resolve(res.data);
      } else {
        reject(e.response.statusText);
      }
    } catch (e) {
      reject("getRating Error: " + e.response.statusText);
    }
  });
}
